src/TgSharp.Core/Network/WSTransport.cs

```diff
@@ -0,0 +1,181 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Reactive.Concurrency;
+using System.Reactive.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using TgSharp.Core.MTProto.Crypto;
+using Websocket.Client;
+using Websocket.Client.Models;
+
+namespace TgSharp.Core.Network
+{
+    public class WSTransport
+    {
+        private WebsocketClient client;
+        private const uint protocol = 0xeeeeeeee;
+        private readonly RandomNumberGenerator rngSource;
+        private readonly string address;
+        private readonly int port;
+        private byte[] encryptKey, encryptIV, decryptKey, decryptIV;
+
+        private byte[] encryptCount = new byte[16], decryptCount = new byte[16];
+        private int encryptNum = 0, decryptNum = 0;
+
+        public Action<Message> OnEncryptedMessage, OnUnencryptedMessage;
+        private static readonly object syncGate = new object();
+
+
+        public WSTransport(string address, int port)
+        {
+            Array.Clear(encryptCount, 0, encryptCount.Length);
+            Array.Clear(decryptCount, 0, decryptCount.Length);
+
+            rngSource = RandomNumberGenerator.Create();
+            this.address = address;
+            this.port = port;
+        }
+
+        private async Task InitializeClient()
+        {
+            if (client != null)
+                client.Dispose();
+
+            client = new WebsocketClient(new Uri($"ws://{address}:{port}/apiws"));
+
+            client
+                .MessageReceived
+                .Select(msg => Observable.FromAsync(async () =>
+                {
+                    await HandleIncomingMessage(msg);
+                }))
+                .Concat()
+                .Subscribe();
+
+            client
+                .ReconnectionHappened
+                .Select(msg => Observable.FromAsync(async () =>
+                {
+                    await ReconnectionHappened(msg);
+                }))
+                .Concat()
+                .Subscribe();
+
+            await client.StartOrFail();
+        }
+
+        private async Task ReconnectionHappened(ReconnectionInfo obj)
```

Comment:
this is a bad name for this method; the name seems to suggest that it returns a bool without doing anything, but it actually does something, it reconnects! let's rename this method